File: src/RadioSim.cpp
/*
 * mm-radio HAL (https://github.com/GloDroid/mm-radio)
 *
 * Copyright (C) 2021 The Android Open Source Project
 * Copyright (C) 2022 GloDroid project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "RadioSim.h"

#include <map>

#include "debug.h"

#define RADIO_MODULE "Sim"

namespace android::hardware::radio::mm {

using ::aidl::android::hardware::radio::RadioIndicationType;
using ::aidl::android::hardware::radio::sim::AppStatus;
using ::ndk::ScopedAStatus;
namespace aidl = ::aidl::android::hardware::radio::sim;
constexpr auto ok = &ScopedAStatus::ok;

ScopedAStatus RadioSim::areUiccApplicationsEnabled(int32_t serial) {
    LOG_STUB << serial;
    mResponse->areUiccApplicationsEnabledResponse(okay(serial), true);
    return ok();
}

ScopedAStatus RadioSim::changeIccPin2ForApp(int32_t serial, const std::string& /*oldPin2*/,
                                            const std::string& /*newPin2*/,
                                            const std::string& /*aid*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->changeIccPin2ForAppResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::changeIccPinForApp(int32_t serial, const std::string& /*oldPin*/,
                                           const std::string& /*newPin*/,
                                           const std::string& /*aid*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->changeIccPinForAppResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::enableUiccApplications(int32_t serial, bool /*enable*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->enableUiccApplicationsResponse(okay(serial));
    return ok();
}

ScopedAStatus RadioSim::getAllowedCarriers(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getAllowedCarriersResponse(notSupported(serial), {}, {});
    return ok();
}

ScopedAStatus RadioSim::getCdmaSubscription(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getCdmaSubscriptionResponse(notSupported(serial), {}, {}, {}, {}, {});
    return ok();
}

ScopedAStatus RadioSim::getCdmaSubscriptionSource(int32_t serial) {
    LOG_STUB << serial;
    mResponse->getCdmaSubscriptionSourceResponse(okay(serial),
                                                 aidl::CdmaSubscriptionSource::RUIM_SIM);
    return ok();
}

ScopedAStatus RadioSim::getFacilityLockForApp(  //
        int32_t serial, const std::string& /*facility*/, const std::string& /*password*/,
        int32_t /*serviceClass*/, const std::string& /*appId*/) {
    LOG_STUB << serial;
    mResponse->getFacilityLockForAppResponse(okay(serial), 0);
    return ok();
}

ScopedAStatus RadioSim::getIccCardStatus(int32_t serial) {
    LOG_STUB << serial;

    auto cardStatus = (aidl::CardStatus){
            .cardState = aidl::CardStatus::STATE_PRESENT,
            .universalPinState = aidl::PinState::UNKNOWN,
            .gsmUmtsSubscriptionAppIndex = 0,
            .cdmaSubscriptionAppIndex = -1,
            .imsSubscriptionAppIndex = -1,
            .atr = {},
            .iccid = "8938003992840681512F",
    };
    cardStatus.applications.emplace_back((AppStatus){
            .appType = AppStatus::APP_TYPE_USIM,
            .appState = AppStatus::APP_STATE_READY,
            .aidPtr = "0xA0 0x00 0x00 0x00 0x87 0x10 0x02 0xFF 0xFF 0xFF 0xFF 0x89",
            .appLabelPtr = "",
            .pin1 = aidl::PinState::UNKNOWN,
            .pin2 = aidl::PinState::UNKNOWN,
    });
    cardStatus.applications.emplace_back((AppStatus){
            .appType = AppStatus::APP_TYPE_RUIM,
    });
    cardStatus.applications.emplace_back((AppStatus){
            .appType = AppStatus::APP_TYPE_ISIM,
    });
    mResponse->getIccCardStatusResponse(okay(serial), cardStatus);
    return ok();
}

ScopedAStatus RadioSim::getImsiForApp(int32_t serial, const std::string& aid) {
    LOG_STUB << serial << "aid: " << aid;
    mResponse->getImsiForAppResponse(okay(serial), "255010899987259");
    return ok();
}

ScopedAStatus RadioSim::getSimPhonebookCapacity(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getSimPhonebookCapacityResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::getSimPhonebookRecords(int32_t serial) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->getSimPhonebookRecordsResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioSim::iccCloseLogicalChannel(int32_t serial, int32_t channelId) {
    LOG_STUB << serial;
    auto resp = okay(serial);
    if (channelId < 1) resp.error = RadioError::INVALID_ARGUMENTS;
    mResponse->iccCloseLogicalChannelResponse(resp);
    return ok();
}

// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
std::map<aidl::IccIo, aidl::IccIoResult> fakeIccProfile = {
        // EF_DIR
        {{0xc0, 0x2f00, "3F00", 0x0, 0x0, 0x0f},
         {144, 0, "621A8205422100300483022F008A01058B032F0601800200C08801F0"}},
        {{0xb2, 0x2f00, "3F00", 0x1, 0x4, 0x30},
         {144, 0,
          "61184F10A0000003431002FF86FF0389FFFFFFFF50044353494DFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
          "FFFFFFFF"}},
        {{0xb2, 0x2f00, "3F00", 0x2, 0x4, 0x30},
         {144, 0,
          "61184F10A0000003431002FF86FF0389FFFFFFFF50044353494DFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
          "FFFFFFFF"}},
        {{0xb2, 0x2f00, "3F00", 0x3, 0x4, 0x30},
         {144, 0,
          "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
          "FFFFFFFF"}},
        {{0xb2, 0x2f00, "3F00", 0x4, 0x4, 0x30},
         {144, 0,
          "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
          "FFFFFFFF"}},
        {{0xc0, 0x2fe2, "3F00", 0x0, 0x0, 0x0f},  // 0xc0=get response 0x2fe2=EF_ICCID
         {144, 0, "000082020000040000000000000000"}},
        {{0xb0, 0x2fe2, "3F00", 0x0, 0x0, 0x0a},  // ]0x2fe2=EF_ICCID
         {144, 0, "9883"}},
        {{0xb0, 0x2fe2, "3F00", 0x0, 0x0, 33282},  //  0x2fe2=EF_ICCID
         {144, 0, "9883"}},
        {{0xc0, 0x2f05, "3F00", 0x0, 0x0, 0x0f},  // EF_PL
         {144, 0, "62178202412183022F058A01058B032F060280020004880128"}},
        {{0xb0, 0x2f05, "3F00", 0x0, 0x0, 0x04}, {144, 0, "FFFFFFFF"}},
        // USIM
        {{0xc0, 0x6f40, "3F007FFF", 0x0, 0x0, 0x0f},  // EF_MSISDN
         {144, 0, "621982054221001C0283026F408A01058B036F0605800200388800"}},
        {{0xb2, 0x6f40, "3F007FFF", 0x1, 0x4, 0x1c},
         {144, 0, "000000000000000000000000000007915155214365F7FFFFFFFFFFFF"}},
        {{0xc0, 0x6fc9, "3F007FFF", 0x0, 0x0, 0x0f},  // EF_MBI
         {144, 0, "62198205422100050183026FC98A01058B036F0602800200058800"}},
        {{0xb2, 0x6fc9, "3F007FFF", 0x1, 0x4, 0x05}, {144, 0, "0100000000"}},
        {{0xc0, 0x6fc7, "3F007FFF", 0x0, 0x0, 0x0f},  // EF_MBDN
         {144, 0, "621982054221001C0283026F408A01058B036F06058002003E8800"}},
        {{0xb2, 0x6fc7, "3F007FFF", 0x1, 0x4, 0x1c},
         {144, 0, "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF07915155674523F1FFFFFFFFFFFF"}},
        {{0xc0, 0x6FAD, "3F007FFF", 0x0, 0x0, 0x0f},  // EF_AD
         {144, 0, "62178202412183026FAD8A01058B036F060180020004880118"}},
        {{0xb2, 0x6FAD, "3F007FFF", 0x0, 0x0, 0x04}, {144, 0, "00000003"}},
        {{0xc0, 0x6FCA, "3F007FFF", 0x0, 0x0, 0x0f},  // EF_MWIS
         {144, 0, "62198205422100050183026FCA8A01058B036F060E800200058800"}},
        {{0xb2, 0x6FCA, "3F007FFF", 0x1, 0x4, 0x05}, {144, 0, "0000000000"}},
        {{0xc0, 0x6F11, "3F007FFF", 0x0, 0x0, 0x0f},  // EF_VOICE_MAIL_INDICATOR_CPHS
         {106, 130, ""}},
        {{0xc0, 0x6F7B, "3F007FFF", 0x0, 0x0, 0x0f},  // EF_FPLMN
         {144, 0, "621C8202412183026F7BA5038001718A01058B036F06038002001E880168"}},
        {{0xb2, 0x6F7B, "3F007FFF", 0x0, 0x0, 0x1e},
         {144, 0, "64F00064F02064F04064F07064F080FFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"}},
        // Other
        {{0xc0, 0x6f3c, "3F007FFF", 0x0, 0x0, 0x0f}, {144, 0, "0000820500000400000000000001B0"}},
        {{0xc0, 0x6f05, "3F007FFF", 0x0, 0x0, 0x0f}, {144, 0, "FFFF"}},
        {{0xc0, 0x6f07, "3F007FFF", 0x0, 0x0, 0x0f}, {144, 0, "0829"}},
        {{0xc0, 0x6f17, "3F007FFF", 0x0, 0x0, 0x0f}, {106, 130, ""}},
        {{0xc0, 0x2f05, "3F00", 0x0, 0x0, 0x0f}, {144, 0, "000082020000040000000000000000"}},
};

ScopedAStatus RadioSim::iccIoForApp(int32_t serial, const aidl::IccIo& iccIo) {
    LOG_STUB << serial;
    auto iccIoKey = (aidl::IccIo){
            .command = iccIo.command,
            .fileId = iccIo.fileId,
            .path = iccIo.path,
            .p1 = iccIo.p1,
            .p2 = iccIo.p2,
            .p3 = iccIo.p3,
    };

    if (fakeIccProfile.count(iccIoKey) != 0) {
        mResponse->iccIoForAppResponse(okay(serial), fakeIccProfile[iccIoKey]);
    } else {
        const aidl::IccIoResult kFileNotFound = {106, 130, ""};
        mResponse->iccIoForAppResponse(okay(serial), kFileNotFound);
    }

    return ok();
}

ScopedAStatus RadioSim::iccOpenLogicalChannel(int32_t serial, const std::string& /*aid*/,
                                              int32_t /*p2*/) {
    LOG_STUB << serial;
    static int channel = 1;
    mResponse->iccOpenLogicalChannelResponse(okay(serial), channel++, {});
    return ok();
}

ScopedAStatus RadioSim::iccTransmitApduBasicChannel(int32_t serial,
                                                    const aidl::SimApdu& /*message*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->iccTransmitApduBasicChannelResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::iccTransmitApduLogicalChannel(int32_t serial,
                                                      const aidl::SimApdu& /*message*/) {
    LOG_STUB << serial;
    mResponse->iccTransmitApduLogicalChannelResponse(okay(serial), aidl::IccIoResult());
    return ok();
}

ScopedAStatus RadioSim::reportStkServiceIsRunning(int32_t serial) {
    LOG_STUB << serial;
    mResponse->reportStkServiceIsRunningResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioSim::requestIccSimAuthentication(  //
        int32_t serial, int32_t /*authContext*/, const std::string& /*authData*/,
        const std::string& /*aid*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->requestIccSimAuthenticationResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::responseAcknowledgement() {
    LOG_UNIMPLEMENTED;
    return ok();
}

ScopedAStatus RadioSim::sendEnvelope(int32_t serial, const std::string& /*command*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->sendEnvelopeResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::sendEnvelopeWithStatus(int32_t serial, const std::string& /*contents*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->sendEnvelopeWithStatusResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::sendTerminalResponseToSim(int32_t serial,
                                                  const std::string& /*commandResponse*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->sendTerminalResponseToSimResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioSim::setAllowedCarriers(  //
        int32_t serial, const aidl::CarrierRestrictions& /*carriers*/,
        aidl::SimLockMultiSimPolicy /*mp*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setAllowedCarriersResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioSim::setCarrierInfoForImsiEncryption(
        int32_t serial, const aidl::ImsiEncryptionInfo& /*imsiEncryptionInfo*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setCarrierInfoForImsiEncryptionResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioSim::setCdmaSubscriptionSource(int32_t serial,
                                                  aidl::CdmaSubscriptionSource /*cdmaSub*/) {
    LOG_STUB << serial;
    mResponse->setCdmaSubscriptionSourceResponse(okay(serial));
    return ok();
}

ScopedAStatus RadioSim::setFacilityLockForApp(  //
        int32_t serial, const std::string& /*facility*/, bool /*lockState*/,
        const std::string& /*password*/, int32_t /*serviceClass*/, const std::string& /*appId*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setFacilityLockForAppResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::setResponseFunctions(
        const std::shared_ptr<aidl::IRadioSimResponse>& response,
        const std::shared_ptr<aidl::IRadioSimIndication>& indication) {
    LOG_STUB << response << ' ' << indication;

    mResponse = response;
    mIndication = indication;

    mIndication->simStatusChanged(RadioIndicationType::UNSOLICITED);
    mIndication->subscriptionStatusChanged(RadioIndicationType::UNSOLICITED, true);

    return ok();
}

ScopedAStatus RadioSim::setSimCardPower(int32_t serial, aidl::CardPowerState /*powerUp*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setSimCardPowerResponse(okay(serial));
    return ok();
}

ScopedAStatus RadioSim::setUiccSubscription(int32_t serial,
                                            const aidl::SelectUiccSub& /*uiccSub*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->setUiccSubscriptionResponse(notSupported(serial));
    return ok();
}

ScopedAStatus RadioSim::supplyIccPin2ForApp(int32_t serial, const std::string& /*pin2*/,
                                            const std::string& /*aid*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->supplyIccPin2ForAppResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::supplyIccPinForApp(int32_t serial, const std::string& /*pin*/,
                                           const std::string& /*aid*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->supplyIccPinForAppResponse(notSupported(serial), {});
    return ok();
}

ScopedAStatus RadioSim::supplyIccPuk2ForApp(int32_t serial, const std::string& /*puk2*/,
                                            const std::string& /*pin2*/,
                                            const std::string& /*aid*/) {
    LOG_UNIMPLEMENTED << serial;
    auto resp = okay(serial);
    resp.error = RadioError::INVALID_SIM_STATE;
    mResponse->supplyIccPuk2ForAppResponse(resp, {});
    return ok();
}

ScopedAStatus RadioSim::supplyIccPukForApp(int32_t serial, const std::string& /*puk*/,
                                           const std::string& /*pin*/, const std::string& /*aid*/) {
    LOG_UNIMPLEMENTED << serial;
    auto resp = okay(serial);
    resp.error = RadioError::INVALID_SIM_STATE;
    mResponse->supplyIccPukForAppResponse(resp, {});
    return ok();
}

ScopedAStatus RadioSim::supplySimDepersonalization(int32_t serial, aidl::PersoSubstate /*pss*/,
                                                   const std::string& /*controlKey*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->supplySimDepersonalizationResponse(notSupported(serial), {}, {});
    return ok();
}

ScopedAStatus RadioSim::updateSimPhonebookRecords(int32_t serial,
                                                  const aidl::PhonebookRecordInfo& /*recordInfo*/) {
    LOG_UNIMPLEMENTED << serial;
    mResponse->updateSimPhonebookRecordsResponse(notSupported(serial), {});
    return ok();
}

}  // namespace android::hardware::radio::mm

File: src/radio/radio_sim.rs
/*
 * mm-radio HAL (https://github.com/GloDroid/mm-radio)
 *
 * SPDX-License-Identifier: GPL-3.0-only
 * Copyright (C) 2023 The GloDroid Project
 */

use crate::{
    mm_zbus::{mm_modem_proxy::ModemProxy, mm_sim_proxy::SimProxy},
    utils::{
        error::Error,
        iradio::{
            declare_async_iradio, def, entry_check, invalid_arg, not_implemented, okay, resp_err,
            shared, sharedmut,
        },
    },
};
use android_hardware_radio::aidl::android::hardware::radio::{
    RadioError::RadioError, RadioIndicationType::RadioIndicationType,
};
use android_hardware_radio_sim::aidl::android::hardware::radio::sim::{
    AppStatus::*, CardPowerState::*, CardStatus::*, CarrierRestrictions::*,
    CdmaSubscriptionSource::*, IRadioSim::*, IRadioSimIndication::*, IRadioSimResponse::*,
    IccIo::*, ImsiEncryptionInfo::*, PersoSubstate::*, PhonebookRecordInfo::*, PinState::*,
    SelectUiccSub::*, SimApdu::*, SimLockMultiSimPolicy::*,
};

use async_std::{sync::RwLock, task::block_on};
use binder::{BinderFeatures, Strong};
use log::info;
use std::sync::Arc;
use zbus::export::async_trait::async_trait;

use super::fake_icc_profile;

#[derive(Default)]
pub struct RadioSimShared {
    modem_bound: bool,
    response: Option<binder::Strong<dyn IRadioSimResponse>>,
    indication: Option<binder::Strong<dyn IRadioSimIndication>>,

    modem_proxy: Option<ModemProxy<'static>>,
    sim_proxy: Option<SimProxy<'static>>,

    uiccapp_enabled: bool,
    card_power_state: CardPowerState,
    channel: i32,
}

#[derive(Default)]
pub struct RadioSim {
    pub shared: Arc<RwLock<RadioSimShared>>,
}

impl RadioSimShared {
    pub(crate) fn bind(
        shared_in: &Arc<RwLock<RadioSimShared>>,
        modem_proxy: &ModemProxy<'static>,
    ) -> Result<(), Error> {
        /* Setup shared structure */
        {
            let mut shared = block_on(shared_in.write());
            shared.modem_bound = true;
            shared.modem_proxy = Some(modem_proxy.clone());
            let conn = modem_proxy.connection();
            let sim_path = block_on(modem_proxy.sim());
            if let Ok(sim_path) = sim_path {
                shared.sim_proxy = Some(block_on(SimProxy::builder(conn).path(sim_path)?.build())?);
            }
        }
        /* Notify framework */
        {
            let shared = block_on(shared_in.read());
            shared.notify_framework()?;
        }
        Ok(())
    }

    pub(crate) fn unbind(shared_in: &Arc<RwLock<RadioSimShared>>) -> Result<(), Error> {
        {
            let mut shared = block_on(shared_in.write());
            shared.modem_proxy = None;
            shared.sim_proxy = None;
            shared.modem_bound = false;
        }
        let shared = block_on(shared_in.read());
        shared.notify_framework()?;
        Ok(())
    }

    fn is_initialized(&self) -> bool {
        self.response.is_some() && self.indication.is_some() && self.modem_proxy.is_some()
    }

    fn notify_framework(&self) -> Result<(), Error> {
        if !self.is_initialized() {
            return Ok(());
        }
        let ind = self.indication.as_ref().ok_or(Error::noneopt())?;
        ind.simStatusChanged(RadioIndicationType::UNSOLICITED)?;
        ind.subscriptionStatusChanged(RadioIndicationType::UNSOLICITED, true)?;
        Ok(())
    }
}

#[async_trait]
impl IRadioSimAsyncServer for RadioSim {
    async fn areUiccApplicationsEnabled(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, areUiccApplicationsEnabledResponse, false);
        okay!(&self, serial, areUiccApplicationsEnabledResponse, shared!(&self).uiccapp_enabled)
    }
    async fn changeIccPin2ForApp(
        &self,
        serial: i32,
        _old_pin2: &str,
        _new_pin2: &str,
        _aid: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, changeIccPin2ForAppResponse, 0)
    }
    async fn changeIccPinForApp(
        &self,
        serial: i32,
        _old_pin: &str,
        _new_pin: &str,
        _aid: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, changeIccPinForAppResponse, 0)
    }
    async fn enableUiccApplications(&self, serial: i32, enable: bool) -> binder::Result<()> {
        entry_check!(&self, serial, enableUiccApplicationsResponse);
        sharedmut!(&self).uiccapp_enabled = enable;
        okay!(&self, serial, enableUiccApplicationsResponse)
    }
    async fn getAllowedCarriers(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getAllowedCarriersResponse, &def(), def())
    }
    async fn getCdmaSubscription(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getCdmaSubscriptionResponse, "", def(), def(), def(), def())
    }
    async fn getCdmaSubscriptionSource(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getCdmaSubscriptionSourceResponse, def());
        okay!(&self, serial, getCdmaSubscriptionSourceResponse, CdmaSubscriptionSource::RUIM_SIM)
    }
    async fn getFacilityLockForApp(
        &self,
        serial: i32,
        _facility: &str,
        _password: &str,
        _service_class: i32,
        _app_id: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, getFacilityLockForAppResponse, 0)
    }
    async fn getIccCardStatus(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, getIccCardStatusResponse, &def());

        let cs: Result<CardStatus, Error> = try {
            let shared = shared!(&self);
            let sim_proxy = shared.sim_proxy.as_ref().ok_or(Error::noneopt())?;
            CardStatus {
                cardState: if sim_proxy.active().await? { STATE_PRESENT } else { STATE_ABSENT },
                universalPinState: PinState::UNKNOWN,
                applications: if shared.card_power_state == CardPowerState::POWER_UP {
                    vec![AppStatus {
                        appType: APP_TYPE_USIM,
                        appState: APP_STATE_READY,
                        pin1: PinState::UNKNOWN,
                        pin2: PinState::UNKNOWN,
                        ..Default::default()
                    }]
                } else {
                    vec![]
                },
                gsmUmtsSubscriptionAppIndex: 0,
                cdmaSubscriptionAppIndex: -1,
                imsSubscriptionAppIndex: -1,
                iccid: sim_proxy.sim_identifier().await?.to_string(),
                eid: sim_proxy.eid().await?,
                ..Default::default()
            }
        };
        let cs = cs?;
        okay!(&self, serial, getIccCardStatusResponse, &cs)
    }
    async fn getImsiForApp(&self, serial: i32, _aid: &str) -> binder::Result<()> {
        entry_check!(&self, serial, getImsiForAppResponse, "");
        let imsi: Result<String, Error> = try {
            let shared = shared!(&self);
            let sim_proxy = shared.sim_proxy.as_ref().ok_or(Error::noneopt())?;
            sim_proxy.imsi().await?
        };
        let imsi = imsi?;
        okay!(&self, serial, getImsiForAppResponse, imsi.as_str())
    }
    async fn getSimPhonebookCapacity(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getSimPhonebookCapacityResponse, &def())
    }
    async fn getSimPhonebookRecords(&self, serial: i32) -> binder::Result<()> {
        not_implemented!(&self, serial, getSimPhonebookRecordsResponse)
    }
    async fn iccCloseLogicalChannel(&self, serial: i32, channel_id: i32) -> binder::Result<()> {
        if channel_id <= 0 {
            return invalid_arg!(&self, serial, iccCloseLogicalChannelResponse);
        }
        entry_check!(&self, serial, iccCloseLogicalChannelResponse);
        okay!(&self, serial, iccCloseLogicalChannelResponse)
    }

    async fn iccIoForApp(&self, serial: i32, icc_io: &IccIo) -> binder::Result<()> {
        entry_check!(&self, serial, iccIoForAppResponse, &def());
        let icc_resp = fake_icc_profile::get_default_sim_io(icc_io);
        info!("ICC Response: {:?}", icc_resp);
        okay!(&self, serial, iccIoForAppResponse, &icc_resp)
    }
    async fn iccOpenLogicalChannel(&self, serial: i32, _aid: &str, _p2: i32) -> binder::Result<()> {
        entry_check!(&self, serial, iccOpenLogicalChannelResponse, 0, &[]);
        sharedmut!(&self).channel += 1;
        okay!(&self, serial, iccOpenLogicalChannelResponse, shared!(&self).channel, &[])
    }
    async fn iccTransmitApduBasicChannel(
        &self,
        serial: i32,
        _message: &SimApdu,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, iccTransmitApduBasicChannelResponse, &def())
    }
    async fn iccTransmitApduLogicalChannel(
        &self,
        serial: i32,
        _message: &SimApdu,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, iccTransmitApduLogicalChannelResponse, &def())
    }
    async fn reportStkServiceIsRunning(&self, serial: i32) -> binder::Result<()> {
        entry_check!(&self, serial, reportStkServiceIsRunningResponse);
        okay!(&self, serial, reportStkServiceIsRunningResponse)
    }
    async fn requestIccSimAuthentication(
        &self,
        serial: i32,
        _auth_context: i32,
        _auth_data: &str,
        _aid: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, requestIccSimAuthenticationResponse, &def())
    }
    async fn responseAcknowledgement(&self) -> binder::Result<()> {
        Ok(())
    }
    async fn sendEnvelope(&self, serial: i32, _contents: &str) -> binder::Result<()> {
        not_implemented!(&self, serial, sendEnvelopeResponse, "")
    }
    async fn sendEnvelopeWithStatus(&self, serial: i32, _contents: &str) -> binder::Result<()> {
        not_implemented!(&self, serial, sendEnvelopeWithStatusResponse, &def())
    }
    async fn sendTerminalResponseToSim(&self, serial: i32, _contents: &str) -> binder::Result<()> {
        not_implemented!(&self, serial, sendTerminalResponseToSimResponse)
    }
    async fn setAllowedCarriers(
        &self,
        serial: i32,
        _carriers: &CarrierRestrictions,
        _multi_sim_policy: SimLockMultiSimPolicy,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setAllowedCarriersResponse)
    }
    async fn setCarrierInfoForImsiEncryption(
        &self,
        serial: i32,
        _imsi_encryption_info: &ImsiEncryptionInfo,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setCarrierInfoForImsiEncryptionResponse)
    }
    async fn setCdmaSubscriptionSource(
        &self,
        serial: i32,
        _cdma_sub: CdmaSubscriptionSource,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setCdmaSubscriptionSourceResponse)
    }
    async fn setFacilityLockForApp(
        &self,
        serial: i32,
        _facility: &str,
        _lock_state: bool,
        _password: &str,
        _service_class: i32,
        _app_id: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setFacilityLockForAppResponse, 0)
    }
    async fn setSimCardPower(&self, serial: i32, power_up: CardPowerState) -> binder::Result<()> {
        entry_check!(&self, serial, setSimCardPowerResponse);
        sharedmut!(&self).card_power_state = power_up;
        okay!(&self, serial, setSimCardPowerResponse)
    }
    async fn setUiccSubscription(
        &self,
        serial: i32,
        _uicc_sub: &SelectUiccSub,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, setUiccSubscriptionResponse)
    }
    async fn supplyIccPin2ForApp(
        &self,
        serial: i32,
        _pin2: &str,
        _aid: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, supplyIccPin2ForAppResponse, 0)
    }
    async fn supplyIccPinForApp(&self, serial: i32, _pin: &str, _aid: &str) -> binder::Result<()> {
        not_implemented!(&self, serial, supplyIccPinForAppResponse, 0)
    }
    async fn supplyIccPuk2ForApp(
        &self,
        serial: i32,
        _puk2: &str,
        _pin2: &str,
        _aid: &str,
    ) -> binder::Result<()> {
        resp_err!(&self, serial, RadioError::PASSWORD_INCORRECT, supplyIccPuk2ForAppResponse, 0)
    }
    async fn supplyIccPukForApp(
        &self,
        serial: i32,
        _puk: &str,
        _pin: &str,
        _aid: &str,
    ) -> binder::Result<()> {
        resp_err!(&self, serial, RadioError::PASSWORD_INCORRECT, supplyIccPukForAppResponse, 0)
    }
    async fn supplySimDepersonalization(
        &self,
        serial: i32,
        _perso_type: PersoSubstate,
        _control_key: &str,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, supplySimDepersonalizationResponse, def(), def())
    }
    async fn updateSimPhonebookRecords(
        &self,
        serial: i32,
        _record_info: &PhonebookRecordInfo,
    ) -> binder::Result<()> {
        not_implemented!(&self, serial, updateSimPhonebookRecordsResponse, 0)
    }
    async fn setResponseFunctions(
        &self,
        radio_response: &binder::Strong<dyn IRadioSimResponse>,
        radio_indication: &binder::Strong<dyn IRadioSimIndication>,
    ) -> binder::Result<()> {
        info!("RadioSim::setResponseFunctions");

        let mut shared = sharedmut!(&self);
        shared.response = Some(radio_response.clone());
        shared.indication = Some(radio_indication.clone());
        shared.uiccapp_enabled = true;
        shared.card_power_state = CardPowerState::POWER_UP;
        shared.notify_framework()?;

        Ok(())
    }
}

declare_async_iradio!(RadioSim, IRadioSim, BnRadioSim);
